static/js/message_list_data.js

```diff
@@ -182,15 +177,37 @@ export class MessageListData {
         });
     }
 
+    messages_filtered_for_user_mutes(messages) {
+        return messages.filter((message) => {
+            if (message.type !== "private") {
+                return true;
+            }
+            const recipients = util.extract_pm_recipients(message.to_user_ids);
+            if (recipients.length > 1) {
+                // Huddle message
+                return true;
+            }
+
+            const recipient_id = Number.parseInt(recipients[0], 10);
+            return !muting.is_user_muted(recipient_id) && !muting.is_user_muted(message.sender_id);
+        });
+    }
+
     unmuted_messages(messages) {
-        return this.messages_filtered_for_topic_mutes(messages);
+        return this.messages_filtered_for_topic_mutes(
+            this.messages_filtered_for_user_mutes(messages),
+        );
     }
 
     update_items_for_topic_muting() {
         if (!this.excludes_muted_topics) {
             return;
         }
-        this._items = this.unmuted_messages(this._all_items);
+        this._items = this.messages_filtered_for_topic_mutes(this._all_items);
+    }
+
+    update_items_for_user_muting() {
+        this._items = this.messages_filtered_for_user_mutes(this._all_items);
```

Comment:
It seems wrong that both of these functions start from `_all_items` and store that to `_items` -- won't that make calling one or the other oscillate between whether topic or user muting is respected?

Reply:
I think the conclusion is we just need an `update_items_for_muting` function that gets run for both updates.

Reply:
> It seems wrong that both of these functions start from _all_items and store that to _items -- won't that make calling one or the other oscillate between whether topic or user muting is respected?

This is a blunder. Really sorry about that. The tests did not catch this because the only tested the behavior of the `MessageListData.unmuted_messages` (which returns filtered messages) and not what is stored in `_items`

Reply:
I've made it so that we now have a single function to update messages after muting events, and also added another assertion to check that `_items` stores the correct set of messages.

Reply:
Sounds good.  Yeah, tests can only do so much: you have to reason through how something will work.  Which is one reason code review is really important: Having two people make sure the logic makes sense to them avoids a lot of potential bugs.